pkg/replication/deployment_replication.go

```diff
@@ -0,0 +1,378 @@
+	"github.com/arangodb/kube-arangodb/pkg/util/retry"
+	"github.com/arangodb/kube-arangodb/pkg/util/trigger"
+)
+
+// Config holds configuration settings for a DeploymentReplication
+type Config struct {
+	Namespace string
+}
+
+// Dependencies holds dependent services for a DeploymentReplication
+type Dependencies struct {
+	Log     zerolog.Logger
+	KubeCli kubernetes.Interface
+	CRCli   versioned.Interface
+}
+
+// deploymentReplicationEvent strongly typed type of event
+type deploymentReplicationEventType string
+
+const (
+	eventArangoDeploymentReplicationUpdated deploymentReplicationEventType = "DeploymentReplicationUpdated"
+)
+
+// seploymentReplicationEvent holds an event passed from the controller to the deployment replication.
+type deploymentReplicationEvent struct {
+	Type                  deploymentReplicationEventType
+	DeploymentReplication *api.ArangoDeploymentReplication
+}
+
+const (
+	deploymentReplicationEventQueueSize = 100
+	minInspectionInterval               = time.Second // Ensure we inspect the generated resources no less than with this interval
+	maxInspectionInterval               = time.Minute // Ensure we inspect the generated resources no less than with this interval
+)
+
+// DeploymentReplication is the in process state of an ArangoDeploymentReplication.
+type DeploymentReplication struct {
+	apiObject *api.ArangoDeploymentReplication // API object
+	status    api.DeploymentReplicationStatus  // Internal status of the CR
+	config    Config
+	deps      Dependencies
+
+	eventCh chan *deploymentReplicationEvent
+	stopCh  chan struct{}
+	stopped int32
+
+	eventsCli corev1.EventInterface
+
+	inspectTrigger         trigger.Trigger
+	recentInspectionErrors int
+	clientCache            client.ClientCache
+}
+
+// New creates a new DeploymentReplication from the given API object.
+func New(config Config, deps Dependencies, apiObject *api.ArangoDeploymentReplication) (*DeploymentReplication, error) {
+	if err := apiObject.Spec.Validate(); err != nil {
+		return nil, maskAny(err)
+	}
+	dr := &DeploymentReplication{
+		apiObject: apiObject,
+		status:    *(apiObject.Status.DeepCopy()),
+		config:    config,
+		deps:      deps,
+		eventCh:   make(chan *deploymentReplicationEvent, deploymentReplicationEventQueueSize),
+		stopCh:    make(chan struct{}),
+		eventsCli: deps.KubeCli.Core().Events(apiObject.GetNamespace()),
+	}
+
+	go dr.run()
+
+	return dr, nil
+}
+
+// Update the deployment replication.
+// This sends an update event in the event queue.
+func (dr *DeploymentReplication) Update(apiObject *api.ArangoDeploymentReplication) {
+	dr.send(&deploymentReplicationEvent{
+		Type: eventArangoDeploymentReplicationUpdated,
+		DeploymentReplication: apiObject,
+	})
+}
+
+// Delete the deployment replication.
+// Called when the local storage was deleted by the user.
+func (dr *DeploymentReplication) Delete() {
+	dr.deps.Log.Info().Msg("deployment replication is deleted by user")
+	if atomic.CompareAndSwapInt32(&dr.stopped, 0, 1) {
```

Comment:
is dr.stopped automatically initialized to 0?

Reply:
Everything is go is initialized to its "zero" value. For ints that is `0`.